nixos/etebase-server: replace customIni with more flexible settings

felschr opened the pull request:
###### Motivation for this change

Fixes incorrect documentation of the `customIni` option reported in #112834.

While having this opportunity I've added the new option `extraConfig` to replace `customIni` which is now merged with the default configuration and still allows overriding options when needed. This makes it easier to add or change single options rather than creating an entirely new config.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/474

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/488

sternenseemann commented on nixos/modules/services/misc/etebase-server.nix:
```diff
@@ -105,30 +114,27 @@ in
-      customIni = mkOption {
-        type = with types; nullOr str;
-        default = null;
+      extraConfig = mkOption {
```
It's better to call this `settings` over `extraConfig` in line with <https://github.com/NixOS/rfcs/pull/42>. We are slowly migrating towards that style of module if possible.

sternenseemann commented on nixos/modules/services/misc/etebase-server.nix:
```diff
@@ -9,30 +9,38 @@ let
 
-  configIni = if cfg.customIni != null then cfg.customIni else defaultConfigIni;
+  configIni = toString (iniFmt.generate "etebase-server.ini"
+    (recursiveUpdate defaultConfigIni cfg.extraConfig));
```
Instead of doing this `recursiveUpdate`, you could use [freeform modules](https://nixos.org/manual/nixos/stable/index.html#sec-freeform-modules) to add supoptions to `extraConfig`/`settings` with preset default values. That would also make the default configuration more transparend to users.

I wonder if it would be possible to get rid of all top-level options of this module in favor of `settings`?

felschr replied:
Thanks for the great feedback. I really like this configuration approach.
I've adapted the module, do you mind taking a look again?

I can't get rid of all top-level options as they're used in other parts of the configuration outside of the ini.

sternenseemann replied:
I think that is a quite considerable reduction in top-level options and the ones that remain there are very reasonable to have top-level!

sternenseemann reviewed (commented):
Just a very quick pass, let me know if you have any questions!

sternenseemann commented on nixos/modules/services/misc/etebase-server.nix:
```diff
@@ -77,58 +66,80 @@ in
+              };
+              media_root = mkOption {
+                type = types.str;
+                default = "${cfg.dataDir}/media";
```
Makes it clearer in docs that this depends on the configuration of the dataDir value.

```suggestion
                default = "${cfg.dataDir}/media";
                defaultText = "\${config.services.etebase-server.dataDir}/media";
```

sternenseemann reviewed (commented):
Overall this looks pretty good to me! I've pointed a few things which could improve usability for this further.

Another thing you could consider if you feel like it, is to add a [NixOS test](https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests) for etebase-server (assuming this is possible at all).

felschr committed:
nixos/etebase-server: replace customIni with more flexible settings option

Fixes #112834 where incorrect documentation of options was reported.

customIni used to replace the entire default configuration while the
new settings option is merged in with the default configuration
and still allows overriding options when needed.

felschr commented:
Alright, I've also added a simple test now. That was actually way easier than I thought 😅.

felschr committed:
nixos/tests/etebase-server: init

felschr committed:
nixos/etebase-server: add missing static directory setup

aanderse reviewed (approved):
This is some fantastic work by @felschr based on some fantastic recommendations and review by @sternenseemann :tada: 

Everyone gets bonus points today! :sparkles:

sternenseemann commented:
@ofborg test etebase-server

sternenseemann commented on nixos/modules/services/misc/etebase-server.nix:
```diff
@@ -8,31 +8,28 @@ let
+    (mkRemovedOptionModule
+      [ "services" "etebase-server" "database" ]
+      "Set the option `services.etebase-server.settings.database' instead.")
+    (mkRenamedOptionModule
```
This doesn't work right since `customIni` used to be of type `str`?!

felschr replied:
Oh, I got a little confused about that option. I changed it back to `mkRemovedOptionModule`.

sternenseemann commented:
@ofborg test etebase-server

sternenseemann commented:
@ofborg test etebase-server
